tripy/examples/diffusion/model.py

```diff
@@ -0,0 +1,775 @@
+# https://arxiv.org/pdf/2112.10752.pdf
+# https://github.com/ekagra-ranjan/huggingface-blog/blob/main/stable_diffusion.md
+# Adapted from https://github.com/tinygrad/tinygrad/blob/master/examples/stable_diffusion.py
+
+import gzip, math, re, os, pathlib, platform, tempfile, hashlib
+import urllib.request
+from functools import lru_cache, reduce
+from tqdm import tqdm
+from collections import namedtuple
+from typing import List, Callable, Optional, Union
+
+import numpy as np
+import tripy as tp
+
+# convenience methods adapted from tinygrad/tensor.py (https://docs.tinygrad.org/tensor/ops/)
+def scaled_dot_product_attention(
+    query: tp.Tensor,
+    key: tp.Tensor,
+    value: tp.Tensor,
+    embedding_dim: Optional[int] = None,
+    attn_mask: Optional[tp.Tensor] = None,
+    is_causal: bool = False,
+) -> tp.Tensor:
+    """
+    Computes scaled dot-product attention.
+    `self` is the query tensor, `key` is the key tensor, and `value` is the value tensor.
+
+    - Described: https://paperswithcode.com/method/scaled
+    - Paper: https://arxiv.org/abs/1706.03762v7
+    """
+
+    if is_causal:  # this path is not called in demoDiffusion
```

Comment:
Can we drop this if it's not called?

Reply:
I have left it for now in case we want to add it as part of the API somewhere